library/x509_csr.c

```diff
@@ -277,12 +279,21 @@ int mbedtls_x509_csr_parse( mbedtls_x509_csr *csr, const unsigned char *buf, siz
     /* Avoid calling mbedtls_pem_read_buffer() on non-null-terminated string */
     if( buf[buflen - 1] == '\0' )
     {
-        mbedtls_pem_init( &pem );
-        ret = mbedtls_pem_read_buffer( &pem,
-                               "-----BEGIN CERTIFICATE REQUEST-----",
-                               "-----END CERTIFICATE REQUEST-----",
-                               buf, NULL, 0, &use_len );
+        cert_header = "-----BEGIN CERTIFICATE REQUEST-----";
+        cert_footer = "-----END CERTIFICATE REQUEST-----";
+        start = strstr( (const char *) buf, cert_header);
```

Comment:
I think we should leave the detection of headers to `mbedtls_pem_read_buffer()`. It has a dedicated error code `MBEDTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT` in case they are not present, so we call `mbedtls_pem_read_buffer()` with `-----{BEGIN/END} CERTIFICATE REQUEST-----` first, and fall back to `----{BEGIN/END} NEW CERTIFICATE REQUEST-----` if it fails with `MBEDTLS_ERR_PEM_NO_HEADER_FOOTER_PRESENT`.